### PR TITLE
Deprecate `tape` and `qtape` properties on the `QNode`

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

The QNode currently has `QNode.tape` and `QNode.qtape` properties, which track the last tape generated during a construct call. Users can check these to verify what was executed most recently.

Now, we’ve added `construct_tape(qnode)(*args, **kwargs)` in the workflow module as a preferred method for constructing tapes. This function creates a tape without modifying the QNode, ensuring it doesn’t interfere with other parts of the codebase.

**Description of the change**

- _Hopefully_ removed all calls to `QNode.tape` and `QNode.qtape` in the PennyLane source code.
- Since many tests rely on these properties to access tape information, I’ve temporarily suppressed the warnings. This needs to be revisited in the backlog and gradually updated to use the preferred code [sc-78317].

Found references in,
- QML: https://github.com/PennyLaneAI/qml/pull/1266
- Qiskit:  https://github.com/PennyLaneAI/pennylane-qiskit/pull/602

No references in,
- Lightning
- Catalyst
- AQT
- IONQ
- Qulacs
- Cirq

**Benefits:** Removes problems and confusion with having a mutable tape property.

**Possible Drawbacks:** None

[sc-76836]

### CODE REVIEW

File: pennylane/gradients/classical_jacobian.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def classical_preprocessing(*args, **kwargs):
         kwargs.pop("shots", None)
         kwargs.pop("argnums", None)
         qnode.construct(args, kwargs)
```
Comment (albi3ro) [inline pennylane/gradients/classical_jacobian.py:144]:
```suggestion
```

File: pennylane/gradients/parameter_shift_hessian.py [hunk 2]
```diff
@@ -522,7 +522,7 @@ def param_shift_hessian(
         results of these tapes into the Hessian:
 
         >>> circuit(x)  # generate the QuantumTape inside the QNode
```
Comment (albi3ro) [inline pennylane/gradients/parameter_shift_hessian.py:524]:
```suggestion
```

File: pennylane/math/multi_dispatch.py [hunk 3]
```diff
@@ -1033,7 +1035,7 @@ def jax_argnums_to_tape_trainable(qnode, argnums, program, args, kwargs):
     ]
 
     qnode.construct(args_jvp, kwargs)
```
Comment (albi3ro) [inline pennylane/math/multi_dispatch.py:1037]:
```suggestion
```

File: pennylane/optimize/qnspsa.py [hunk 4]
```diff
@@ -370,9 +370,11 @@ def _get_spsa_grad_tapes(self, cost, args, kwargs):
             args_minus[index] = arg - self.finite_diff_step * direction
 
         cost.construct(args_plus, kwargs)
```
Comment (albi3ro) [inline pennylane/optimize/qnspsa.py:372]:
```suggestion
```

File: pennylane/optimize/qnspsa.py [hunk 5]
```diff
@@ -370,9 +370,11 @@ def _get_spsa_grad_tapes(self, cost, args, kwargs):
...
-        tape_plus = cost.tape.copy(copy_operations=True)
+        tape = qml.workflow.construct_tape(cost)(*args_plus, **kwargs)
+        tape_plus = tape.copy(copy_operations=True)
         cost.construct(args_minus, kwargs)
```
Comment (albi3ro) [inline pennylane/optimize/qnspsa.py:375]:
```suggestion
```

File: pennylane/optimize/qnspsa.py [hunk 6]
```diff
@@ -425,22 +427,26 @@ def _get_overlap_tape(self, cost, args1, args2, kwargs):
...
 
     @staticmethod
     def _get_operations(cost, args, kwargs):
         cost.construct(args, kwargs)
-        return cost.tape.operations
+        tape = qml.workflow.construct_tape(cost)(*args, **kwargs)
+        return tape.operations
 
     def _apply_blocking(self, cost, args, kwargs, params_next):
         cost.construct(args, kwargs)
```
Comment (albi3ro) [inline pennylane/optimize/qnspsa.py:440]:
```suggestion
```

File: pennylane/optimize/qnspsa.py [hunk 7]
```diff
@@ -425,22 +427,26 @@ def _get_overlap_tape(self, cost, args1, args2, kwargs):
...
-        tape_loss_curr = cost.tape.copy(copy_operations=True)
+        tape = qml.workflow.construct_tape(cost)(*args, **kwargs)
+        tape_loss_curr = tape.copy(copy_operations=True)
 
         if not isinstance(params_next, list):
             params_next = [params_next]
 
         cost.construct(params_next, kwargs)
```
Comment (albi3ro) [inline pennylane/optimize/qnspsa.py:447]:
```suggestion
```

File: pennylane/optimize/shot_adaptive.py [hunk 8]
```diff
@@ -309,7 +309,7 @@ def _single_shot_qnode_gradients(self, qnode, args, kwargs):
         self.check_device(qnode.device)
 
         qnode.construct(args, kwargs)
```
Comment (albi3ro) [inline pennylane/optimize/shot_adaptive.py:311]:
```suggestion
```

File: pennylane/transforms/core/transform_program.py [hunk 9]
```diff
@@ -387,7 +387,7 @@ def classical_preprocessing(program, *args, **kwargs):
             kwargs.pop("shots", None)
             kwargs.pop("argnums", None)
             qnode.construct(args, kwargs)
```
Comment (albi3ro) [inline pennylane/transforms/core/transform_program.py:389]:
```suggestion
```

File: pennylane/transforms/core/transform_program.py [hunk 10]
```diff
@@ -457,7 +457,8 @@ def _jacobian(*args, **kwargs):
                     classical_preprocessing, sub_program, argnums, *args, **kwargs
                 )
                 qnode.construct(args, kwargs)
```
Comment (albi3ro) [inline pennylane/transforms/core/transform_program.py:459]:
```suggestion
```

File: tests/test_qnode.py [hunk 11]
```diff
@@ -458,22 +473,24 @@ def func(x, y):
         y = pnp.array(0.54, requires_grad=True)
 
         res = qn(x, y)
+        tape = qml.workflow.construct_tape(qn)(x, y)
```
Comment (albi3ro) [inline tests/test_qnode.py:476]: I'd actually leave this test case as is, but put `with pytest.warns(qml.PennyLaneDeprecationWarning)` around accessing the tapes.  Then we can delete this test when we remove `QNode.tape`. 

This test seems to be explicitly about the `QNode.tape` proprety, so want to make sure we are deleting it when we delete `QNode.tape`.

File: tests/test_qnode.py [hunk 12]
```diff
@@ -621,21 +638,23 @@ def func(x, y):
...
-    assert len(func.qtape.operations) == 3
-    assert len(func.qtape.observables) == 1
-    assert func.qtape.num_params == 2
+    assert isinstance(tape, QuantumScript)
+    assert len(tape.operations) == 3
+    assert len(tape.observables) == 1
+    assert tape.num_params == 2
 
-    expected = qml.execute([func.tape], dev, None)
+    expected = qml.execute([tape], dev, None)
```
Comment (albi3ro) [inline tests/test_qnode.py:648]:
```suggestion
    expected = np.cos(x)
```

File: tests/test_qnode.py [hunk 13]
```diff
@@ -621,21 +638,23 @@ def func(x, y):
...
 
     # when called, a new quantum tape is constructed
-    old_tape = func.qtape
+    old_tape = tape
     res2 = func(x, y)
+    new_tape = qml.workflow.construct_tape(func)(x, y)
 
     assert np.allclose(res, res2, atol=tol, rtol=0)
-    assert func.qtape is not old_tape
+    assert new_tape is not old_tape
```
Comment (albi3ro) [inline tests/test_qnode.py:657]: Again, I think we can just delete the tape checking.  We know it's giving the correct answers, so that's what we should care about.

**Review by albi3ro — APPROVED** [review]
❤️

File: doc/releases/changelog-dev.md [hunk 14]
```diff
@@ -98,6 +98,10 @@
 
 <h3>Deprecations 👋</h3>
 
+* The ``tape`` and ``qtape`` properties of ``QNode`` have been deprecated. 
+  Instead, use the ``qml.workflow.construct_tape`` function.
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:102]:
```suggestion
* The `tape` and `qtape` properties of `QNode` have been deprecated. 
  Instead, use the `qml.workflow.construct_tape` function.
```

File: tests/test_qnode_legacy.py [hunk 15]
```diff
@@ -1096,19 +1117,22 @@ def circuit(a, shots=0):
...
-        assert circuit.qtape.operations[0].wires.labels == (0,)
+        tape = qml.workflow.construct_tape(circuit)(0.8, shots=0)
+        assert tape.operations[0].wires.labels == (0,)
 
     # pylint: disable=unexpected-keyword-arg
     def test_no_shots_per_call_if_user_has_shots_qfunc_arg(self):
         """Tests that the per-call shots overwriting is suspended
         if user has a shots argument, but a warning is raised."""
-        dev = qml.device("default.mixed", wires=[0, 1], shots=10)
+        dev = qml.device("default.qubit", wires=[0, 1], shots=10)
```
Comment (mudit2812) [inline tests/test_qnode_legacy.py:1135]: This file tests legacy devices, hence why `default.mixed` is used. Now that `default.mixed` is being ported over, we might need a different device. Maybe we should use the `tests/default_qubit_legacy` device? cc @albi3ro

File: tests/circuit_graph/test_circuit_graph.py [hunk 16]
```diff
@@ -29,6 +30,13 @@
 from pennylane.wires import Wires
 
 
+@pytest.fixture(autouse=True)
+def suppress_tape_property_deprecation_warning():
+    warnings.filterwarnings(
+        "ignore", "The tape/qtape property is deprecated", category=qml.PennyLaneDeprecationWarning
+    )
```
Comment (mudit2812) [inline tests/circuit_graph/test_circuit_graph.py:37]: Why is warning suppression needed in so many files?

**Review by mudit2812 — COMMENTED** [review]
Looks good. There are a couple of comments that are blocking approval from me.

File: tests/circuit_graph/test_circuit_graph.py (hunk 16 above)
Comment (andrijapau) [reply tests/circuit_graph/test_circuit_graph.py:37]: All these files rely on either `tape` or `qtape` to access the QNode’s tape for performing checks. Addressing every instance in this PR (over 200 occurrences) would be impractical. Instead, this will be an ongoing effort.

Comment (mudit2812) [reply tests/circuit_graph/test_circuit_graph.py:37]: Thanks, resolving.

File: tests/test_qnode_legacy.py (hunk 15 above)
Comment (andrijapau) [reply tests/test_qnode_legacy.py:1135]: After speaking with @albi3ro, I think it will be addressed in a follow-up PR. 😄

**Review by mudit2812 — APPROVED** [review]
